### DISCUSSION TITLE
split teal compiler to multi modules

### DISCUSSION THREAD

**Discussion by virusdefender** [teal-language/tl#394] [General]
`tl.tl` is tooo long, may be we can split it to multi modules, like `lexer` `parser` `std_types` and so on.

It's not easy for new users(for example, me) to contribute to the project.

**Comment by bew** [comment]
This is exactly an issue I was going to open today ^^ with a possible solution!
I'll write down my thoughts, and it can be moved to a separate issue later if this is something the project wants to go to

What I'm thinking about is like you said, splitting the main code in multiple modules, but for distribution it's not good, it is simpler to have a single file like `tl.lua` to put in your app or library.

**single file gen**:
To solve this, I think `tl gen` command could have a `--single-file` option, that takes all the required tl modules (and lua ones?) and put them in a single lua file for distribution.

**tl distributions**:
Then it could allow making different distributions of tl, including various sets of modules.
So the core distributed tl can remain small.

For example the normal distribution would include main tl modules, then another compiler-debug distribution could include additional tools to debug issues with the compiler.
Another use case would be to have a distribution including classes systems based on metaprogramming (as described in one of the meetup), but this whole thing would be opt-in by using the tl-classes distribution of tl, etc...
(example invocations: `tl gen --single-module tl-core.tl -o tl.lua` or `tl gen --single-module tl-for-debug.tl -o tl-debug.lua`)

**Reply by lenscas** [reply]
> To solve this, I think tl gen command could have a --single-file option, that takes all the required tl modules (and lua ones?) and put them in a single lua file for distribution.

There is an issue about this already <https://github.com/teal-language/tl/issues/67>
